### ISSUE TITLE
(suggestion) sf::Packet need a "reserve" method

### ISSUE THREAD

**Issue by JonathSpirit** [SFML/SFML#1752]
class sf::Packet need a reserve method.

It can be really great to have a sf::Packet::reserve(size_t n) method to optimize big workload on packet.

Currently, any call to operator << (or append) will trigger a reallocation. This can be improve if the devlopper know exactly the size we want to append when having multiple variables to packed.

A simple implementation will be like that : 
`
void Packet::reserve(std::size_t sizeInBytes)
{
        m_data.reserve(sizeInBytes);
}
`

This is simply a suggestion, please tell me what you think

**Comment by Bromeon** [comment]
> Currently, any call to operator << (or append) will trigger a reallocation. 

Not exactly. `sf::Packet` uses an underlying `std::vector`, which has an exponential growth strategy. This means adding an element is amortized O(1). 

`reserve()` would only be helpful if you create lots of packets, and for each of them you add very small elements (thus rather `<<` than `append()`). However, there's a much better way to migitate this: reuse the `sf::Packet` instance with `clear()`. Not only will this avoid reallocations (the buffer doesn't shrink), but it will also make the initial allocation unnecessary (which `reserve()` still requires).

Unless you experienced real performance issues with `sf::Packet`, I would say the above-mentioned usage gives better optimization results -- I would thus advocate to keep the API simple.

**Comment by 1aam2am1** [comment]
O would say that keeping and reusing packet is hard. In my program as it is now i can create packet, then place my data and send it. 

But with reusing i would need static packet. As many thread could use my function i would need some mutex to guard my static variable or create static local thread variable so that each thread in this function has his own packet. Now there is second problem i should check all code that this function use so that there would not be any recursion as this would corupt my data.

Another idea for reusing packet would be to create shared packet in factory and then when we have used it completely releasing it to factory or checking in factory each time when we need new one if some shared packet has one use_cound meaning only factory has pointer to it.

Third idea is too keep packet as member of some class as this class will use this packed for sending data. It don't have as many problems as above, but it can't be used everywhere. And it will hoar memory as long as it exists. Example: Program sends data with size of 10kb. And one time it send 1gb file. Now our packet will have vector with 2gb declared space as long as it exists. Even when we need only 10kb for running.

Each and every idea how to reuse packet is somewhat wrong or problematic.

The best idea would be to declare how much wee will need. 

Allocations in vector are in 2*existing size as mentioned above. But wee need to copy existing data. While this takes away time.

Push => Allock size (1) copy(0)
Push => Allock size (2) copy(1)
Push => Allock size (4) copy(2)
Push => size (4)
Push => Allock size (8) copy(4)
Push x3 => size (8)
Push => Allock size (16) copy(8)

As wee can see in the beginning wee have many allocations that are unnecessary. We could hardcore reserve(64) in constructor or << operator in packet too make it somewhat more bearable.

This function should be implemented with mark that it can be deleted/removed any time. In this way when something changes so that this function will be no longer possible or necessary we can remove it. As users where warned that this pard of code is experimental and can change anytime. In this way prices of making sfml better would be faster as we could get better response from users, what they think of new functions and Api.

**Comment by Bromeon** [comment]
> But with reusing i would need static packet. As many thread could use my function i would need some mutex to guard my static variable or create static local thread variable so that each thread in this function has his own packet. 

Obviously don't do that. I think we are overstating the problem of dynamic allocations a bit -- other programming languages do it all the time. What I'm saying is that if you truly face the issue that the few growth allocations are exactly the problem, then you might as well optimize away the initial one by reusing it. Not that this is the generally recommended way of using `sf::Packet`.

> Third idea is too keep packet as member of some class as this class will use this packed for sending data. It don't have as many problems as above, but it can't be used everywhere. And it will hoar memory as long as it exists. Example: Program sends data with size of 10kb. And one time it send 1gb file. Now our packet will have vector with 2gb declared space as long as it exists. Even when we need only 10kb for running.

This is a problem, and also shows why `reserve()` might not be enough but additionally need a `shrink_to_fit()` counterpart. We basically add another layer to the simple data serialization API, and that layer is purely "influence the underlying memory management". We don't do this in other buffer-like APIs, such as `sf::VertexArray` or `sf::ConvexShape`. Not even `sf::String` has it, which is literally a character container.

Also, something we didn't mention yet: it's possible to add the entire data in one go using `append()`. Like this you don't need any reallocations. Of course you can no longer benefit from `operator<<`, but with such advanced needs, chances are you use a different protocol already.
 
> We could hardcore reserve(64) in constructor or << operator in packet too make it somewhat more bearable.

I don't think this is a good idea for the same reason that `std::vector` does not pre-allocate: you waste space if you never use it. It's assuming one specific use case.

> This function should be implemented with mark that it can be deleted/removed any time. In this way when something changes so that this function will be no longer possible or necessary we can remove it. As users where warned that this pard of code is experimental and can change anytime. In this way prices of making sfml better would be faster as we could get better response from users, what they think of new functions and Api.

The thing is, when we add method, we commit to support it for the entire major version, to avoid breaking changes at users. So if we figure out that `std::vector` is not an ideal container internally and switch to something different, we may not be able to support `reserve()` any longer. This is one of the reasons why I think some abstraction is reasonable here.

But otherwise you mention a good point: I really like the concept of experimental vs. stable APIs. Kotlin for example has [`@OptIn`](https://kotlinlang.org/docs/opt-in-requirements.html#non-propagating-opt-in), the Rust language has [nightly/unstable features](https://doc.rust-lang.org/book/appendix-07-nightly-rust.html). Unfortunately I'm not aware of a similar concept in C++; only documenting a symbol will still cause people to use it without realizing its experimental status.

---

**TLDR:** I understand that `reserve()` can have benefits, but I'm not convinced it solves an issue the users actually face, and `sf::Packet` allocations is _the_ bottleneck of their applications. API design is about trade-offs: we want people to have a simple way to use the library, with reasonable performance. There would be a lot of places where we _could_ give extra control to how the internal memory is managed, but we don't do it because a) it decreases the mental overhead and keeps APIs concise about what they functionally offer, and b) it leaves us freedom to change implementation without breaking assumptions about the internals.

**Comment by shane-tran-whitmire** [comment]
Is it possible to have a custom packet type that can do this functionality? For vertex arrays, you can just have your own std::vector<sfml::Vertex>. Can you do the same with packets? If you can, I personally don't think it is a necessary addition to the API.
